File: backend/src/models/UploadJob.ts
import { Model, DataTypes, Optional } from 'sequelize';
import sequelize from '../config/database';
import User from './User';

// UploadJob attributes interface
export interface UploadJobAttributes {
  id: string; // UUID
  user_id: number;
  filename: string;
  file_type: string;
  status: 'pending' | 'processing' | 'completed' | 'failed' | 'cancelled';
  total_records: number;
  new_records: number;
  updated_records: number;
  error_records: number;
  created_at: Date;
  updated_at: Date;
  completed_at?: Date;
}

// Attributes for UploadJob creation - id, timestamps, and default values are optional
export interface UploadJobCreationAttributes extends Optional<UploadJobAttributes, 
  'id' | 'created_at' | 'updated_at' | 'completed_at' | 
  'total_records' | 'new_records' | 'updated_records' | 'error_records'> {}

class UploadJob extends Model<UploadJobAttributes, UploadJobCreationAttributes> implements UploadJobAttributes {
  public id!: string;
  public user_id!: number;
  public filename!: string;
  public file_type!: string;
  public status!: 'pending' | 'processing' | 'completed' | 'failed' | 'cancelled';
  public total_records!: number;
  public new_records!: number;
  public updated_records!: number;
  public error_records!: number;
  public created_at!: Date;
  public updated_at!: Date;
  public completed_at?: Date;

  // Virtual field for calculating processing time
  public get processingTime(): number | null {
    if (!this.completed_at) return null;
    return Math.floor((this.completed_at.getTime() - this.created_at.getTime()) / 1000); // in seconds
  }

  // Virtual field for calculating success rate
  public get successRate(): number {
    if (this.total_records === 0) return 0;
    return ((this.new_records + this.updated_records) / this.total_records) * 100;
  }

  // Method to update job progress
  public async updateProgress(stats: {
    totalRecords?: number;
    newRecords?: number;
    updatedRecords?: number;
    errorRecords?: number;
  }): Promise<UploadJob> {
    if (stats.totalRecords !== undefined) this.total_records = stats.totalRecords;
    if (stats.newRecords !== undefined) this.new_records = stats.newRecords;
    if (stats.updatedRecords !== undefined) this.updated_records = stats.updatedRecords;
    if (stats.errorRecords !== undefined) this.error_records = stats.errorRecords;
    
    await this.save();
    return this;
  }

  // Method to mark job as completed
  public async markAsCompleted(): Promise<UploadJob> {
    this.status = 'completed';
    this.completed_at = new Date();
    await this.save();
    return this;
  }

  // Method to mark job as failed
  public async markAsFailed(): Promise<UploadJob> {
    this.status = 'failed';
    this.completed_at = new Date();
    await this.save();
    return this;
  }

  // Method to cancel the job
  public async cancel(): Promise<UploadJob> {
    this.status = 'cancelled';
    this.completed_at = new Date();
    await this.save();
    return this;
  }
}

UploadJob.init(
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    user_id: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: 'users',
        key: 'id',
      },
    },
    filename: {
      type: DataTypes.STRING(255),
      allowNull: false,
      validate: {
        notEmpty: true,
      },
    },
    file_type: {
      type: DataTypes.STRING(10),
      allowNull: false,
      validate: {
        isIn: [['csv', 'xlsx']],
      },
    },
    status: {
      type: DataTypes.STRING(20),
      allowNull: false,
      defaultValue: 'pending',
      validate: {
        isIn: [['pending', 'processing', 'completed', 'failed', 'cancelled']],
      },
    },
    total_records: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 0,
    },
    new_records: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 0,
    },
    updated_records: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 0,
    },
    error_records: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 0,
    },
    created_at: {
      type: DataTypes.DATE,
      allowNull: false,
      defaultValue: DataTypes.NOW,
    },
    updated_at: {
      type: DataTypes.DATE,
      allowNull: false,
      defaultValue: DataTypes.NOW,
    },
    completed_at: {
      type: DataTypes.DATE,
      allowNull: true,
    },
  },
  {
    sequelize,
    modelName: 'UploadJob',
    tableName: 'upload_jobs',
    timestamps: false, // We'll manually manage created_at and updated_at
    hooks: {
      beforeUpdate: (uploadJob: UploadJob) => {
        uploadJob.updated_at = new Date();
      },
    },
  }
);

// Set up the association with the User model
UploadJob.belongsTo(User, { foreignKey: 'user_id', as: 'user' });

export default UploadJob;

File: flask_backend/app/models/upload_job.py
from app import db
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.sql import func # For default timestamps
import uuid # For UUID default
from datetime import datetime

class UploadJob(db.Model):
    __tablename__ = 'upload_jobs'

    # Using String for ID to store UUID, with default from Python's uuid
    id = db.Column(db.String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    filename = db.Column(db.String(255), nullable=False)
    file_type = db.Column(db.String(10), nullable=False) # e.g., 'csv', 'xlsx'
    status = db.Column(db.String(20), nullable=False, default='pending') # 'pending', 'processing', 'completed', 'failed', 'cancelled'
    total_records = db.Column(db.Integer, default=0)
    processed_records = db.Column(db.Integer, default=0) # Added from init.sql
    new_records = db.Column(db.Integer, default=0)
    updated_records = db.Column(db.Integer, default=0)
    error_records = db.Column(db.Integer, default=0)
    error_details = db.Column(db.Text, nullable=True) # Added from init.sql
    started_at = db.Column(db.DateTime(timezone=True), nullable=True) # Added from init.sql
    completed_at = db.Column(db.DateTime(timezone=True), nullable=True)
    created_at = db.Column(db.DateTime(timezone=True), server_default=func.now())
    updated_at = db.Column(db.DateTime(timezone=True), server_default=func.now(), onupdate=func.now())


    __table_args__ = (
        db.Index('idx_upload_jobs_status', 'status'),
    )

    def __repr__(self):
        return f'<UploadJob {self.id} [{self.status}]>'

    # Camel case representation for API responses
    def to_dict(self):
        return {
            "id": self.id,
            "filename": self.filename,
            "fileType": self.file_type,
            "status": self.status,
            "totalRecords": self.total_records,
            "processedRecords": self.processed_records,
            "newRecords": self.new_records,
            "updatedRecords": self.updated_records,
            "errorRecords": self.error_records,
            "errorDetails": self.error_details,
            "startedAt": self.started_at.isoformat() if self.started_at else None,
            "completedAt": self.completed_at.isoformat() if self.completed_at else None,
            "createdAt": self.created_at.isoformat() if self.created_at else None,
            "updatedAt": self.updated_at.isoformat() if self.updated_at else None,
        }
